File: src/travelpost/writers/map/js/PostIcon.js
if (typeof L.Travel === "undefined") {
  L.Travel = {};
}

L.Travel.PostIcon = L.Icon.extend({
  options: {
    imgUrl: null,
    iconShape: "circle", // "rounded-square", "square"
    emptySize: 16,
    imgSize: 32,

    backgroundColor: "#3388ff",
    borderColor: "white",
    borderWidth: 2,
    borderStyle: "solid",
    className: "",
  },

  initialize: function (options) {
    L.Util.setOptions(this, options);
    const size = this.options.imgUrl
      ? this.options.imgSize
      : this.options.emptySize;

    this.options.iconSize = [size, size];
    this.options.iconAnchor = [
      this.options.iconSize[0] / 2,
      this.options.iconSize[1] / 2 - 1,
    ];
    this.options.popupAnchor = [0, this.options.iconSize[1] / 2];
    this.options.tooltipAnchor = [0, -this.options.iconSize[1] / 2];
  },

  createIcon: function (oldIcon) {
    const div =
      oldIcon && oldIcon.tagName === "DIV"
        ? oldIcon
        : document.createElement("div");

    div.innerHTML = "";

    this._setIconStyles(div);

    return div;
  },

  createShadow: function () {
    return null;
  },

  _setIconStyles: function (iconDiv) {
    const options = this.options,
      size = L.point(options.iconSize),
      anchor = L.point(options.iconAnchor);

    iconDiv.className = "leaflet-marker-icon post-icon";
    if (options.iconShape) {
      iconDiv.className += " " + options.iconShape;
    }
    if (options.className) {
      iconDiv.className += " " + options.className;
    }
    if (options.imgUrl) {
      iconDiv.style.backgroundImage = `url('${options.imgUrl}')`;
    } else {
      iconDiv.style.backgroundColor = options.backgroundColor;
    }
    if (options.borderColor && options.borderWidth) {
      iconDiv.style.borderColor = options.borderColor;
      iconDiv.style.borderStyle = options.borderStyle;
      iconDiv.style.borderWidth = options.borderWidth + "px";
    }

    if (anchor) {
      iconDiv.style.marginLeft = -anchor.x + "px";
      iconDiv.style.marginTop = -anchor.y + "px";
    }

    if (size) {
      iconDiv.style.width = size.x + "px";
      iconDiv.style.height = size.y + "px";
    }
  },
});

L.Travel.postIcon = function (options) {
  return new L.Travel.PostIcon(options);
};

File: src/travelpost/writers/map/post_icon.py
"""Post Icon."""

import folium
from folium.template import Template
from folium.utilities import remove_empty


class PostIcon(folium.MacroElement):
    """Post Icon."""

    _template = Template(
        """
        {% macro script(this, kwargs) %}
            var {{ this.get_name() }} = L.Travel.postIcon({{ this.options|tojavascript }});
        {% endmacro %}
        """  # noqa: E501
    )

    def __init__(
        self,
        img_url: str | None,
        icon_shape: str = "circle",
        icon_size: int = 32,
        empty_size: int = 16,
        background_color: str = "#3388ff",
        border_color: str = "white",
        border_width: int = 2,
        border_style: str = "solid",
        class_name: str | None = None,
    ) -> None:
        super().__init__()
        self._name = "PostIcon"
        self.options = remove_empty(
            img_url=img_url,
            icon_shape=icon_shape,
            icon_size=icon_size,
            empty_size=empty_size,
            background_color=background_color,
            border_color=border_color,
            border_width=border_width,
            border_style=border_style,
            class_name=class_name,
        )
